### PR TITLE
added piccolo provider 

### PR DESCRIPTION
**Pull request by leeeunkoo**
init piccolo

### CODE REVIEW

File: api/Dockerfile_backup [hunk 1]
```diff
@@ -0,0 +1,45 @@
+##
```
Comment (Haishi2016) [inline api/Dockerfile_backup:1]: Please remove the _backup file

File: api/Dockerfile [hunk 2]
```diff
@@ -4,8 +4,8 @@
 ## SPDX-License-Identifier: MIT
 ##
 
-ARG BUILD_BASE_IMAGE=golang:1.22.3-alpine
-ARG TARGET_BASE_IMAGE=ubuntu:latest
+ARG BUILD_BASE_IMAGE=golang:1.22.4-alpine
+ARG TARGET_BASE_IMAGE=ubuntu:24.04
```
Comment (Haishi2016) [inline api/Dockerfile:8]: Is there a specific reason to use this version?

File: api/Dockerfile [hunk 3]
```diff
@@ -37,9 +37,8 @@ RUN \
 ADD https://github.com/golang/go/raw/master/lib/time/zoneinfo.zip /zoneinfo.zip
 ENV ZONEINFO /zoneinfo.zip
 COPY --from=build /dist /
-ADD ./api/symphony-api.json /
-EXPOSE 8080
-EXPOSE 8081
-ENV LOG_LEVEL Debug
-# ENV CONFIG /symphony-api.json
-CMD sh -c 'if [ -f /etc/pki/ca-trust/source/anchors/proxy-cert.crt ]; then update-ca-trust; fi && exec /symphony-api -c $CONFIG -l $LOG_LEVEL'
+ADD ./api/symphony-api-no-k8s.json /
```
Comment (Haishi2016) [inline api/Dockerfile:40]: If this is for your local test environment please don't check this in.

File: api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo_test.go [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+func TestInitWithMap(t *testing.T) {
+	configMap := map[string]string{
+		"name": "name",
+	}
+	provider := PiccoloTargetProvider{}
+	err := provider.InitWithMap(configMap)
+	assert.Nil(t, err)
+}
+func TestPiccoloTargetProviderInstall(t *testing.T) {
+	testPiccoloProvider := os.Getenv("TEST_PICCOLO_PROVIDER")
```
Comment (Haishi2016) [inline api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo_test.go:42]: How can we properly test this from our side? Do you have a mock server that we can use?

File: api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+}
+
+func (i *PiccoloTargetProvider) Apply(ctx context.Context, deployment model.DeploymentSpec, step model.DeploymentStep, isDryRun bool) (map[string]model.ComponentResultSpec, error) {
+	ctx, span := observability.StartSpan("Piccolo Target Provider", ctx, &map[string]string{
+		"method": "Apply",
+	})
+	var err error = nil
+	defer observ_utils.CloseSpanWithError(span, &err)
+
+	sLog.Infof("  P (Piccolo Target): applying artifacts: %s - %s, traceId: %s", deployment.Instance.Spec.Scope, deployment.Instance.ObjectMeta.Name, span.SpanContext().TraceID().String())
```
Comment (msftcoderdjw) [inline api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go:133]: Use sLog.InfofCtx(ctx, xxx) to always log the context.

File: api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+	injections := &model.ValueInjections{
+		InstanceId: deployment.Instance.ObjectMeta.Name,
+		SolutionId: deployment.Instance.Spec.Solution,
+		TargetId:   deployment.ActiveTarget,
+	}
+
+	components := step.GetComponents()
+	err = i.GetValidationRule(ctx).Validate(components)
+	if err != nil {
+		sLog.Errorf("  P (Piccolo Target): failed to validate components: %+v, traceId: %s", err, span.SpanContext().TraceID().String())
```
Comment (msftcoderdjw) [inline api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go:144]: sLog.ErrorfCtx

File: api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+	for _, component := range step.Components {
+		if component.Action == model.ComponentUpdate {
+			name := model.ReadPropertyCompat(component.Component.Properties, "workload.name", injections)
+			if name == "" {
+				err = errors.New("component doesn't have workload.name property")
+				ret[component.Component.Name] = model.ComponentResultSpec{
+					Status:  v1alpha2.UpdateFailed,
+					Message: err.Error(),
+				}
+				sLog.Errorf("  P (Piccolo Target): %+v, traceId: %s", err, span.SpanContext().TraceID().String())
```
Comment (msftcoderdjw) [inline api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go:163]: sLog.ErrorfCtx

Comment (msftcoderdjw) [inline api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go:163]: sLog.ErrorfCtx

File: api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go (hunk 5 above)
Comment (msftcoderdjw) [inline api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go:133]: Use sLog.InfofCtx(ctx, xxx) to always log the context.

File: api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+				return ret, err
+			}
+			resp, err := http.Post("http://0.0.0.0:9090/create-scenario/"+component.Component.Name, "text/plain", nil)
+			if err != nil {
+				sLog.Errorf("  P (Piccolo Target): fail to create resource")
```
Comment (msftcoderdjw) [inline api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go:168]: sLog.ErrorfCtx

File: api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+	})
+	var err error = nil
+	defer observ_utils.CloseSpanWithError(span, &err)
+
+	sLog.Infof("  P (Piccolo Target): getting artifacts: %s - %s, traceId: %s", deployment.Instance.Spec.Scope, deployment.Instance.ObjectMeta.Name, span.SpanContext().TraceID().String())
+
+	ret := make([]model.ComponentSpec, 0)
+	for _, component := range references {
+		name := component.Component.Name
+		req, err := http.NewRequest("GET", "http://0.0.0.0:9090/scenario/"+name, nil)
```
[the remark this answers is not in the document]
Comment (youngtaekiim) [reply api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go:96]: A new commit has been created for the fix.
The PR will be updated soon.

File: api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+		if err != nil {
+			continue
```
[the remark this answers is not in the document]
Comment (youngtaekiim) [reply api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go:98]: A new commit has been created for the fix.
The PR will be updated soon.

File: api/Dockerfile_backup (hunk 1 above)
Comment (youngtaekiim) [reply api/Dockerfile_backup:1]: This was a personal test environment setting.
I created a commit for deletion and the PR will be updated.

File: api/Dockerfile (hunk 2 above)
Comment (youngtaekiim) [reply api/Dockerfile:8]: There are no reason. (personal preference)
This was a personal test environment setting.
I created a commit for deletion and the PR will be updated.

File: api/Dockerfile (hunk 3 above)
Comment (youngtaekiim) [reply api/Dockerfile:40]: This was a personal test environment setting.
I created a commit for deletion and the PR will be updated.

File: api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go (hunk 7 above)
Comment (youngtaekiim) [reply api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go:163]: This commit was written too long ago, so the ErrorfCtx method does not exist.
I will create a new PR later to fix this.

File: api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go (hunk 5 above)
Comment (youngtaekiim) [reply api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go:133]: This commit was written too long ago, so the InfofCtx method does not exist.
I will create a new PR later to fix this.

File: api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go (hunk 8 above)
Comment (youngtaekiim) [reply api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo.go:168]: This commit was written too long ago, so the ErrorfCtx method does not exist.
I will create a new PR later to fix this.

File: api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo_test.go (hunk 4 above)
Comment (youngtaekiim) [reply api/pkg/apis/v1alpha1/providers/target/piccolo/piccolo_test.go:42]: I am not ready for unit testing at the moment.
As you mentioned, I plan to prepare a container image that can act as a mock server for unit testing.

**Pull request closed without merging**